### PR TITLE
Better Handle Dead Accounts in Balances

### PR DESCRIPTION
**Pull request by shawntabrizi**
Before this PR, certain balance functions tried to mutate a dead account causing it to appear in storage. However, those operations on a dead account should be noop.

This does a simple check that the account trying to be modified already exists, if not, it is a noop.

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]
LGTM

**Review by thiolliere — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -1146,9 +1147,10 @@ impl<T: Config<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I
 
 	/// Unreserve some funds, returning any amount that was unable to be unreserved.
 	///
-	/// Is a no-op if the value to be unreserved is zero.
+	/// Is a no-op if the value to be unreserved is zero or the account does not exist.
 	fn unreserve(who: &T::AccountId, value: Self::Balance) -> Self::Balance {
 		if value.is_zero() { return Zero::zero() }
+		if Self::is_dead_account(&who) { return value }
```
Comment (pepyakin) [inline frame/balances/src/lib.rs:1153]: Why don't we do this check in reserve as well?

Comment (shawntabrizi) [reply frame/balances/src/lib.rs:1153]: @pepyakin reserve returns a DispatchResult and will already be a noop if there are not enough funds to reserve. If the value is zero it is also a noop.

Comment (shawntabrizi) [reply frame/balances/src/lib.rs:1153]: The functions to worry about were the ones that use `mutate_account` whereas reserve uses `try_mutate_account`... but maybe you see something I dont?

Comment (pepyakin) [reply frame/balances/src/lib.rs:1153]: nah, that was a genuine question

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**